### DISCUSSION TITLE
What LLMs do you want Pieces to add?

### DISCUSSION THREAD

**Discussion by rosie-at-pieces** [pieces-app/support#121] [Ideas]
We try to add new LLMs to select for your copilot's runtime every time we release, and we receive feedback occasionally requesting specific ones like Mistral, Claude, etc. Let us know below what LLMs you'd love to see in Pieces and you just might get it in an upcoming release!

**Comment by QDaed** [comment]
Some really good LLM Model i've test 
https://huggingface.co/bigcode/starcoder2-15b
https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1 ( not Mistral , it is a better upgrade from Mistral )

**Reply by rosie-at-pieces** [reply]
Mixtral coming soon!!

**Comment by ypark9** [comment]
I would love to see Claude models (Opus, Sonnet, Haiku)!

**Reply by QDaed** [reply]
Voting this also , expect using this with my api for Opus

**Reply by hra42** [reply]
That would be awesome, I now Claude 3 is super expensive, but maybe the pricing could be avoided when you can bring your own api key, 

That would also be awesome as an ad for the upcomming AI service.

**Reply by Calendy** [reply]
claude 3.5 that is the best AI for coding. I don't know how it is not already added.

**Reply by hra42** [reply]
> claude 3.5 that is the best AI for coding. I don't know how it is not already added. @vdripcommunity 

I would bet that it is a cost problem

**Reply by rosie-at-pieces** [reply]
Good news - we have Claude 3 and 3.5 available now! Released last night so make sure to update!

**Reply by hra42** [reply]
> Good news - we have Claude 3 and 3.5 available now! Released last night so make sure to update!  

Great news, works fine! I just would update the UI to actually show model you use (Claude 3/3.5 Sonnet), btw. I'd rather have haiku included then the normal 3 sonnet, so you have the fastest cloud llm available at your hands.

**Reply by rosie-at-pieces** [reply]
Good feedback @hra42 - we're in discussions about that right now!

**Reply by rosie-at-pieces** [reply]
And now the rest of the Claude models are in 😎

**Reply by eusd-server-engineer** [reply]
Llama 3.1 8b would be great, or LM Studio/Ollama support.

**Comment by hra42** [comment]
DBRX (https://www.databricks.com/blog/introducing-dbrx-new-state-art-open-llm)

![DBRXStats](https://github.com/pieces-app/support/assets/44985923/ea4cfe79-2d4c-4532-824e-4b72ceb5de5a)

**Reply by eumel1968** [reply]
That would be awesome!

**Comment by jayaveer** [comment]
By utilizing LiteLLM or a similar tool, it becomes possible to quickly onboard more cloud-based and on-device Language Model (LLM) models. The best part is that LiteLLM has direct integration with Hugging Face and Ollama, which means that by using LiteLLM, one can easily support all LLMs that are supported by Hugging Face and Ollama, with minimal effort.

**Reply by CallMeLaNN** [reply]
Agreed. This is the first thing I search after looking into this app. I already have Ollama running, and it already using my storage and memory. I don't want to double waste it.

New and better models come out every a few weeks, I got a couple of better local LLMs then Mistral 7B. So I just use this with phi2 and waiting for this option.

**Comment by rosie-at-pieces** [comment]
Heard Grok today!

**Comment by rosie-at-pieces** [comment]
Also heard Microsoft copilot

**Reply by hra42** [reply]
How about phi-3?

**Reply by eumel1968** [reply]
No not Microsoft Copilot! I have install the Pieces Suite for Developers to avoid Microsoft Copilot.

**Reply by eumel1968** [reply]
It would be nicer to extend Pieces to replace Microsoft Copilot! In this case, you would only need one AI. 

Just an idea!

**Reply by hra42** [reply]
Phi-3 is a local model, you don't need Microsoft Copilot to run it!

**Reply by eumel1968** [reply]
I am sorry. Just realised I replied under the wrong post.

My three posts refer to rosie's post about Microsoft Copilot.

Can someone move this posts?

**Reply by hra42** [reply]
@eumel1968 
no we can't move posts, you need to delete them and post them again

**Comment by csaroff** [comment]
llama 3?

**Reply by icnocop** [reply]
`Llama 3 8B` is now included in Pieces, in the `On-Device` tab.

**Comment by rosie-at-pieces** [comment]
Just heard from a user to add Gemini advanced, and maybe even use your own account credentials for that

**Comment by Horschig** [comment]
Please take a look at CodeQwen1.5, see here:

https://qwenlm.github.io/blog/codeqwen1.5/

It has some very powerful, low parameter models, which outperforms most other LLMs in several benchmarks, eg HumanEval 83% vs GPT-4 Turbo 85%. I haven't used it a lot myself, but so far it's good.

**Reply by eumel1968** [reply]
That would be awesome. I agree with you.

**Comment by Horschig** [comment]
Two news potential interesting ones are wavecoder by Microsoft and granite by IBM. Benchmarks look good, and both companies are somewhat trustable.

**Reply by Horschig** [reply]
![image](https://github.com/pieces-app/support/assets/3985329/082f43e8-b72c-4c42-b841-0ea6d6ac6007)

**Reply by hra42** [reply]
do you have benchmarks for wavecoder by microsoft?

**Reply by Horschig** [reply]
![image](https://github.com/pieces-app/support/assets/3985329/ddd95774-ccd2-48c1-a185-ab36c76705a1)

Link to paper: https://arxiv.org/html/2312.14187v3

**Reply by hra42** [reply]
do you know any external serivce where you can test both models via an api?

**Reply by Horschig** [reply]
Nope, sorry, I know just know how to test them locally via ollama and the like, but I have not searched for any api service.

**Comment by hra42** [comment]
Llama 3.1 (8b, 70b and 405b)

Benchmarks from Meta AI:
![image](https://github.com/user-attachments/assets/10cce941-b2ba-41dc-b2e7-fbddd3555e8a)
![image](https://github.com/user-attachments/assets/c5d7d02a-ce99-480d-9888-62a9647b6fbd)

**Comment by clairefro** [comment]
not an LLM per say, but [Groq](https://groq.com/) as a cloud option for fast inference would be great for productivity.

They have an [API](https://console.groq.com/docs/api-reference#chat-create). The models used by Groq are common open source ones (llama, mistral etc)

**Comment by clairefro** [comment]
Also, an option to point to local models already downloaded on-device would be amazing. 

Right now Ollama users have to re-download large models (4Gb+) to work in Pieces

**Reply by rosie-at-pieces** [reply]
@brian-pieces !

**Reply by brian-at-pieces** [reply]
We're in the process of switching our local LLM backbone to Ollama, so this UX should be much better for you very soon!

**Comment by aryan2090** [comment]
Also, can you route the GPT-4o model to `gpt-4o-2024-08-06` or `chatgpt-4o-latest`.

**Comment by asadkhaksar1122** [comment]
Please add the blackbox ai in this

**Comment by Horschig** [comment]
Relating back to https://github.com/pieces-app/support/discussions/121#discussioncomment-9346765, please add Qwen 2.5 Coder, see https://ollama.com/library/qwen2.5-coder

![image](https://github.com/user-attachments/assets/867ca7ed-4cdf-4e24-aca6-7cfa91b41dbd)

**Reply by brian-at-pieces** [reply]
This is absolutely on our radar. We're finishing up a big on-device LLM update and will have support for Qwen 2.5 Coder and many other new models shortly after!

**Reply by Horschig** [reply]
![image](https://github.com/user-attachments/assets/f2f7fb15-73b0-4722-a27b-0f2bf5313a4d)

**Reply by fabienb** [reply]
Hi. 
Considering I already have Qwen 2.5 Coder 32B Instruct installed locally, I'd love to see it implemented in Pieces as well but perhaps rather than having 18Gb of this model duplicated on my drive, you could allow us a third option: to specify a folder where to find on-device LLMs? Is that doable?

**Comment by onkanat** [comment]
Pieces' Ollama works fine.
I installed curl with qwen2.5-coder etc.
and I use it with another chat user interface.
How do the models I installed in Pieces Desktop work?

**Reply by Horschig** [reply]
the shortest answer is that it is not possible, because the Pieces team is using own (fine-tuned?) versions of the LLMs, and it does not work with other models right out of the box.

**Reply by brian-at-pieces** [reply]
We're actually using the official models from the Ollama registry. Nothing is fine-tuned by us

**Reply by Horschig** [reply]
Then it's a bit weird that we can't use all available models. Is there a reason why?

**Reply by brian-at-pieces** [reply]
Totally agree. We have some infrastructure left over from our initial implementation of local LLMs in the app that's limiting us on this front. We're looking to get rid of this soon and then all Ollama models will be available for use. This was also our first release using Ollama and we wanted to keep the scope fairly small

**Reply by Horschig** [reply]
I understand, thanks for the update. Hope it will be possible soon (together with specifying the ollama endpoint, i.e. ip and port, we could finally start using it within the company ;) ).

**Comment by brian-at-pieces** [comment]
@onkanat we're in the process of adding support for qwen2.5-coder and many other models - this should be out in the next release or two.

In general, when you download and Ollama model via Pieces it will download the q4_K_S variant of the model to your global Ollama model registry, so you can use it in Pieces or any other chat interface that supports this model. We're looking to support all quantizations at some point in the future, but we chose to start with just q4_K_S for now because it's a good middle ground for most of our users.

**Reply by hra42** [reply]
> @onkanat we're in the process of adding support for qwen2.5-coder and many other models - this should be out in the next release or two.
> 
> 
> 
> In general, when you download and Ollama model via Pieces it will download the q4_K_S variant of the model to your global Ollama model registry, so you can use it in Pieces or any other chat interface that supports this model. We're looking to support all quantizations at some point in the future, but we chose to start with just q4_K_S for now because it's a good middle ground for most of our users.

Hey @brian-at-pieces, short question about that, would it be possible to integrate the ollama extension for let's say a custom endpoint, as long as it complies with the ollama spec?

For example this could enable the usage of openrouter or any other local tool that supports the ollama api.

**Reply by brian-at-pieces** [reply]
This is technically possible, but he haven't implemeted it as a public endpoint yet. I'll add it to our backlog. Thanks for the input!

**Reply by projx** [reply]
LS Studio and JAN AI,.. if you can integrate  with these it opens up a whole world of different potential.. another similar "model orchestrator" is MSTY.

**Comment by arthur-t-w-mills** [comment]
I'd love to see Lumo from Proton as Cloud LLM option: https://lumo.proton.me/about

This would be great for users that are more privacy focused. I've been using Lumo for a couple of weeks now and I have to say I'm pretty impressed with the results.
